zookeeper-specifications/zab-1.0/ZabWithFLE.tla

```diff
@@ -0,0 +1,1020 @@
+               replyOk == /\ infoOk
+                          /\ NullPoint \in ackeRecv[i]
+           IN /\ infoOk
+              /\ \/ /\ replyOk
+                    /\ subRECOVERYSYNC(i, j)
+                    /\ ackeRecv' = [ackeRecv EXCEPT ![i] = IF j \notin ackeRecv[i] THEN ackeRecv[i] \union {j}
+                                                                                   ELSE ackeRecv[i]]
+                    /\ UNCHANGED <<state, currentEpoch,logicalClock, receiveVotes, outOfElection, recvQueue, waitNotmsg, leadingVoteSet, electionMsgs, idTable,
+                                   zabState, leaderAddr, learners>>
+                 \/ /\ ~replyOk
+                    /\ \/ /\ logOk
+                          /\ ackeRecv' = [ackeRecv EXCEPT ![i] = IF j \notin ackeRecv[i] THEN ackeRecv[i] \union {j}
+                                                                                         ELSE ackeRecv[i]]
+                          /\ Discard(j, i)
+                          /\ UNCHANGED <<varsL, zabState, leaderAddr, learners, forwarding>>
+                       \/ /\ ~logOk            \* go back to election
+                          /\ LeaderShutdown(i)
+                          /\ UNCHANGED ackeRecv
+                    /\ UNCHANGED <<history, commitIndex, synced, forwarding, ackIndex, committedCounter, proposalMsgsLog>>
+        /\ UNCHANGED <<acceptedEpoch, cepochRecv, ackldRecv,currentCounter, sendCounter, committedIndex, tempVarsZ, cepochSent, epochLeader, inherentViolated>>
+                          
+LeaderDiscovery2(i) ==
+        /\ state[i] = LEADING
+        /\ zabState[i] = DISCOVERY
+        /\ ackeRecv[i] \in Quorums
+        /\ zabState'       = [zabState       EXCEPT ![i] = SYNCHRONIZATION]
+        /\ currentEpoch'   = [currentEpoch   EXCEPT ![i] = acceptedEpoch[i]]
+        /\ initialHistory' = [initialHistory EXCEPT ![i] = history[i]]
+        /\ ackeRecv'       = [ackeRecv       EXCEPT ![i] = ackeRecv[i] \union {NullPoint}]
+        /\ ackIndex'       = [ackIndex       EXCEPT ![i][i] = Len(history[i])]
+        /\ UpdateProposal(i, i, lastZxid[i], currentEpoch'[i])
+        /\ LET epoch == acceptedEpoch[i]
+           IN epochLeader' = [epochLeader EXCEPT ![epoch] = epochLeader[epoch] \union {i}]
+        /\ UNCHANGED <<state, lastZxid, acceptedEpoch, history, commitIndex, logicalClock, receiveVotes, outOfElection, recvQueue, waitNotmsg,
+                       leadingVoteSet, learners, cepochRecv, ackldRecv, forwarding, currentCounter, sendCounter, committedIndex, committedCounter,
+                       tempMaxEpoch, followerVarsZ, proposalMsgsLog, inherentViolated, msgVarsZ, idTable>>
+
+(* Note:  Set cepochRecv, ackeRecv, ackldRecv to {NullPoint} in corresponding three actions to
+          make sure that the prospective leader will not broadcast NEWEPOCH/NEWLEADER/COMMITLD twice.*)
+-----------------------------------------------------------------------------
+RECOVERYSYNC(i, j) ==
```

Comment:
Consider in the future replacing this magic atomic synchronization with DIFF based message passing. The current form could hide a defect. It would increase the state space a lot though, so other state space reduction might be worth doing first.